File: TestApp/android/app/src/main/java/com/testapp/TestAppNativeModule.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.testapp;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.microsoft.appcenter.crashes.model.TestCrashException;
import com.microsoft.appcenter.reactnative.shared.AppCenterReactNativeShared;

import java.util.concurrent.atomic.AtomicInteger;

import javax.annotation.Nonnull;

public class TestAppNativeModule extends ReactContextBaseJavaModule {

    private static final String TEST_APP_NATIVE = "TestAppNative";

    static {
        System.loadLibrary("native-lib");
    }

    private static final String APP_SECRET = "app_secret";

    private static final String START_AUTOMATICALLY = "start_automatically";

    private final SharedPreferences mSharedPreferences;

    private native void nativeAllocateLargeBuffer();

    TestAppNativeModule(ReactApplicationContext context) {
        super(context);
        mSharedPreferences = context.getSharedPreferences(getName(), Context.MODE_PRIVATE);
    }

    static void initSecrets(Context context) {
        SharedPreferences sharedPreferences = context.getSharedPreferences(TEST_APP_NATIVE, Context.MODE_PRIVATE);
        String secretOverride = sharedPreferences.getString(APP_SECRET, null);
        AppCenterReactNativeShared.setAppSecret(secretOverride);
        boolean startAutomaticallyOverride = sharedPreferences.getBoolean(START_AUTOMATICALLY, true);
        AppCenterReactNativeShared.setStartAutomatically(startAutomaticallyOverride);
    }

    @Override
    @Nonnull
    public String getName() {
        return TEST_APP_NATIVE;
    }

    @SuppressWarnings("unused")
    @ReactMethod
    public void configureStartup(String secretString, boolean startAutomatically) {

        /* We need to use empty string in Android for no app secret like in JSON file. */
        if (secretString == null) {
            secretString = "";
        }
        mSharedPreferences.edit()
                .putString(APP_SECRET, secretString)
                .putBoolean(START_AUTOMATICALLY, startAutomatically)
                .apply();
    }

    @SuppressWarnings("unused")
    @ReactMethod
    public void generateTestCrash() {

        /*
         * To crash the test app even in release.
         * We reach this code if Crashes.generateTestCrash detected release mode.
         * We can tell with stack trace whether we used SDK method in debug or this one in release.
         */
        throw new TestCrashException();
    }

    @SuppressWarnings("unused")
    @ReactMethod
    public void produceLowMemoryWarning() {
        final AtomicInteger i = new AtomicInteger(0);
        final Handler handler = new Handler(Looper.getMainLooper());
        handler.post(new Runnable() {
            @Override
            public void run() {
                nativeAllocateLargeBuffer();
                Log.d("TestApp", "Memory allocated: " + i.addAndGet(128) + "MB");
                handler.post(this);
            }
        });
    }
}

File: TestApp/android/app/src/main/java/com/testapp/TestAppNativeModule.kt
package com.testapp

import android.content.Context
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.facebook.react.bridge.BaseJavaModule
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactMethod
import com.microsoft.appcenter.crashes.model.TestCrashException
import com.microsoft.appcenter.reactnative.shared.AppCenterReactNativeShared
import com.microsoft.appcenter.analytics.Analytics
import java.util.concurrent.atomic.AtomicInteger

class TestAppNativeModule(context: Context) : BaseJavaModule() {

    companion object {
        private const val DEMO_APP_NATIVE = "TestAppNative"
        private const val APP_SECRET = "app_secret"
        private const val START_AUTOMATICALLY = "start_automatically"
        private const val MANUAL_SESSION_TRACKER_ENABLED_KEY = "manual_session_tracker_enabled"

        init {
            System.loadLibrary("native-lib")
        }

        fun initSecrets(context: Context) {
            val sharedPreferences = context.getSharedPreferences(DEMO_APP_NATIVE, Context.MODE_PRIVATE)
            val secretOverride = sharedPreferences.getString(APP_SECRET, null)
            AppCenterReactNativeShared.setAppSecret(secretOverride)
            val startAutomaticallyOverride = sharedPreferences.getBoolean(START_AUTOMATICALLY, true)
            AppCenterReactNativeShared.setStartAutomatically(startAutomaticallyOverride)
        }

        fun initManualSessionTrackerState(context: Context) {
            val sharedPreferences = context.getSharedPreferences(DEMO_APP_NATIVE, Context.MODE_PRIVATE)
            val isManualSessionTrackerEnabled = sharedPreferences.getBoolean(MANUAL_SESSION_TRACKER_ENABLED_KEY, false)
            if (isManualSessionTrackerEnabled) {
                Analytics.enableManualSessionTracker()
            }
        }
    }

    private val sharedPreferences: SharedPreferences = 
        context.getSharedPreferences(name, Context.MODE_PRIVATE)

    private external fun nativeAllocateLargeBuffer()

    override fun getName(): String {
        return DEMO_APP_NATIVE
    }

    @ReactMethod
    fun saveManualSessionTrackerState(state: Boolean) {
        sharedPreferences.edit()
            .putBoolean(MANUAL_SESSION_TRACKER_ENABLED_KEY, state)
            .apply()
    }

    @ReactMethod
    fun getManualSessionTrackerState(promise: Promise) {
        promise.resolve(if (sharedPreferences.getBoolean(MANUAL_SESSION_TRACKER_ENABLED_KEY, false)) 1 else 0)
    }

    @ReactMethod
    fun configureStartup(secretString: String?, startAutomatically: Boolean) {
        sharedPreferences.edit()
            .putString(APP_SECRET, secretString ?: "")
            .putBoolean(START_AUTOMATICALLY, startAutomatically)
            .apply()
    }

    @ReactMethod
    fun generateTestCrash() {
        throw TestCrashException()
    }

    @ReactMethod
    fun produceLowMemoryWarning() {
        val atomicInteger = AtomicInteger(0)
        val handler = Handler(Looper.getMainLooper())
        handler.post(object : Runnable {
            override fun run() {
                nativeAllocateLargeBuffer()
                Log.d("TestApp", "Memory allocated: ${atomicInteger.addAndGet(128)}MB")
                handler.post(this)
            }
        })
    }
}
